# 総合演習問題 - 3

スポーツチームの選手の能力を測定したデータがあります．

| 選手 | 判断力 | 協調力 | 俊敏力 | 技術力 | 持久力 |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| p01 | 7.08 | 7.82 | 7.21 | 7.23 | 7.00 |
| p02 | 8.00 | 8.33 | 6.41 | 6.90 | 6.36 |
| p03 | 7.17 | 7.18 | 7.08 | 6.70 | 6.89 |
| p04 | 7.56 | 7.34 | 6.38 | 6.45 | 6.11 |
| p05 | 6.75 | 6.35 | 6.71 | 7.05 | 6.53 |
| p06 | 7.02 | 7.79 | 6.32 | 6.02 | 6.04 |
| p07 | 7.96 | 8.02 | 5.87 | 5.75 | 5.36 |
| p08 | 7.49 | 7.22 | 5.92 | 6.21 | 5.80 |
| p09 | 7.48 | 7.58 | 5.81 | 5.77 | 5.75 |
| p10 | 6.45 | 6.21 | 6.82 | 6.37 | 6.20 |
| p11 | 6.24 | 7.03 | 5.78 | 6.20 | 5.95 |
| p12 | 6.86 | 5.87 | 6.03 | 6.28 | 6.03 |
| p13 | 6.16 | 6.96 | 6.18 | 5.84 | 5.77 |
| p14 | 6.06 | 6.35 | 5.85 | 6.14 | 5.59 |
| p15 | 6.59 | 6.25 | 5.68 | 5.74 | 5.72 |
| p16 | 5.49 | 5.62 | 6.39 | 6.10 | 5.96 |
| p17 | 6.16 | 6.43 | 5.48 | 5.59 | 5.74 |
| p18 | 6.11 | 5.79 | 5.71 | 5.97 | 5.37 |
| p19 | 5.92 | 5.97 | 4.69 | 4.83 | 4.65 |
| p20 | 4.73 | 5.60 | 4.90 | 5.08 | 4.72 |

このデータを主成分の2軸に縮約したデータを，線形代数の対角化理論を用いて求めよ．
計算についてはPythonで行ってください．

#### ヒント

分散共分散行列を求めます．
分散共分散行列は，対称行列になっているので，その固有値は正の実数になります．
よって，分散共分散行列を対角化することにより，固有値と固有ベクトルを求めます．

選手の能力のデータは次の記述をコピーしてご使用ください．

```Python
competency = np.array([
    [7.08,8.00,7.17,7.56,6.75,7.02,7.96,7.49,7.48,6.45,6.24,6.86,6.16,6.06,6.59,5.49,6.16,6.11,5.92,4.73],
    [7.82,8.33,7.18,7.34,6.35,7.79,8.02,7.22,7.58,6.21,7.03,5.87,6.96,6.35,6.25,5.62,6.43,5.79,5.97,5.60],
    [7.21,6.41,7.08,6.38,6.71,6.32,5.87,5.92,5.81,6.82,5.78,6.03,6.18,5.85,5.68,6.39,5.48,5.71,4.69,4.90],
    [7.23,6.90,6.70,6.45,7.05,6.02,5.75,6.21,5.77,6.37,6.20,6.28,5.84,6.14,5.74,6.10,5.59,5.97,4.83,5.08],
    [7.00,6.36,6.89,6.11,6.53,6.04,5.36,5.80,5.75,6.20,5.95,6.03,5.77,5.59,5.72,5.96,5.74,5.37,4.65,4.72]])
```

*****
## 解答

まず最初にNumPyライブラリーをインポートします．

In [1]:
import numpy as np

選手の能力のデータをnumpyの配列として定義します．

In [2]:
competency = np.array([
    [7.08,8.00,7.17,7.56,6.75,7.02,7.96,7.49,7.48,6.45,6.24,6.86,6.16,6.06,6.59,5.49,6.16,6.11,5.92,4.73],
    [7.82,8.33,7.18,7.34,6.35,7.79,8.02,7.22,7.58,6.21,7.03,5.87,6.96,6.35,6.25,5.62,6.43,5.79,5.97,5.60],
    [7.21,6.41,7.08,6.38,6.71,6.32,5.87,5.92,5.81,6.82,5.78,6.03,6.18,5.85,5.68,6.39,5.48,5.71,4.69,4.90],
    [7.23,6.90,6.70,6.45,7.05,6.02,5.75,6.21,5.77,6.37,6.20,6.28,5.84,6.14,5.74,6.10,5.59,5.97,4.83,5.08],
    [7.00,6.36,6.89,6.11,6.53,6.04,5.36,5.80,5.75,6.20,5.95,6.03,5.77,5.59,5.72,5.96,5.74,5.37,4.65,4.72]])

分散共分散行列をnumpy.cov(competency,bias=True)によって求めます．

In [3]:
Cov = np.cov(competency,bias=True)
Cov

array([[ 0.665004  ,  0.558538  ,  0.236881  ,  0.244626  ,  0.234127  ],
       [ 0.558538  ,  0.68718475,  0.2147045 ,  0.2042345 ,  0.2127265 ],
       [ 0.236881  ,  0.2147045 ,  0.392309  ,  0.328349  ,  0.340543  ],
       [ 0.244626  ,  0.2042345 ,  0.328349  ,  0.339689  ,  0.311778  ],
       [ 0.234127  ,  0.2127265 ,  0.340543  ,  0.311778  ,  0.334781  ]])

分散共分散行列の固有値と固有ベクトルをnumpy.linalg.eig()関数によって求めます．

In [4]:
eigen_values,eigen_vectors = np.linalg.eig(Cov)

In [5]:
eigen_values

array([ 1.68393113,  0.5644221 ,  0.11671932,  0.03483811,  0.01905709])

固有値および寄与率，累積寄与率をまとめると次表になります．
固有値の並び順を降順にしています．

|  | 固有値 | 寄与率 | 累積寄与率 |
|:---:|:---:|:---:|:---:|
| $\lambda_1$ | 1.684 | 69.6 % | 69.6 % |
| $\lambda_2$ | 0.564 | 23.3 % | 92.9 % |
| $\lambda_3$ | 0.117 |  4.8 % | 97.8 % |
| $\lambda_4$ | 0.035 |  1.4 % | 99.2 % |
| $\lambda_5$ | 0.019 |  0.9 % | 100 % |


In [6]:
eigen_vectors

array([[ 0.55338237,  0.3939371 ,  0.724545  ,  0.11441119,  0.02293762],
       [ 0.54163078,  0.4987323 , -0.67089838, -0.08146401, -0.03407599],
       [ 0.37749405, -0.48551445, -0.10938394,  0.63083706, -0.4602757 ],
       [ 0.35860585, -0.42099166,  0.08563058, -0.76182108, -0.32628808],
       [ 0.35958885, -0.42792685, -0.07504976,  0.04412411,  0.82461813]])

第1主成分である固有ベクトルは，eigen_vectors[:,0]となります．

In [7]:
eigen_vectors[:,0]

array([ 0.55338237,  0.54163078,  0.37749405,  0.35860585,  0.35958885])

各選手の第1主成分の値は，eigen_vectors[:,0].dot(competency)となります．

In [8]:
eigen_vectors[:,0].dot(competency)

array([ 15.9850742 ,  16.11994564,  15.40954482,  15.07764826,
        14.58395794,  14.82053425,  14.95407244,  14.60273059,
        14.57489349,  14.02112297,  13.80559589,  13.67223034,
        13.68058466,  13.21313438,  13.29139418,  12.82486639,
        13.02883539,  12.74456855,  11.68416087,  11.01932889])

第2主成分である固有ベクトルは，eigen_vectors[:,1]となります．

In [9]:
eigen_vectors[:,1]

array([ 0.3939371 ,  0.4987323 , -0.48551445, -0.42099166, -0.42792685])

各選手の第2主成分の値は，eigen_vectors[:,1].dot(competency)となります．

In [10]:
eigen_vectors[:,1].dot(competency)

array([-2.85065553, -1.43266793, -2.80107545, -1.78875184, -3.19412992,
       -1.53693618, -0.42878736, -1.41914335, -0.98349983, -3.00804997,
       -1.99833095, -2.52191151, -2.03037913, -2.26305044, -1.90883338,
       -3.25534024, -1.83676142, -2.28895916, -0.99077284, -1.88124977])

各選手の能力データについて，第1主成分と第2主成分によって近似したデータは，

```Python
np.round( eigen_vectors[:,0].dot(competency)*eigen_vectors[:,0:1] + \
          eigen_vectors[:,1].dot(competency)*eigen_vectors[:,1:2] ,2).T
```

によって求められます．

In [11]:
np.round( eigen_vectors[:,0].dot(competency)*eigen_vectors[:,0:1] + \
          eigen_vectors[:,1].dot(competency)*eigen_vectors[:,1:2] ,2).T

array([[ 7.72,  7.24,  7.42,  6.93,  6.97],
       [ 8.36,  8.02,  6.78,  6.38,  6.41],
       [ 7.42,  6.95,  7.18,  6.71,  6.74],
       [ 7.64,  7.27,  6.56,  6.16,  6.19],
       [ 6.81,  6.31,  7.06,  6.57,  6.61],
       [ 7.6 ,  7.26,  6.34,  5.96,  5.99],
       [ 8.11,  7.89,  5.85,  5.54,  5.56],
       [ 7.52,  7.2 ,  6.2 ,  5.83,  5.86],
       [ 7.68,  7.4 ,  5.98,  5.64,  5.66],
       [ 6.57,  6.09,  6.75,  6.29,  6.33],
       [ 6.85,  6.48,  6.18,  5.79,  5.82],
       [ 6.57,  6.15,  6.39,  5.96,  6.  ],
       [ 6.77,  6.4 ,  6.15,  5.76,  5.79],
       [ 6.42,  6.03,  6.09,  5.69,  5.72],
       [ 6.6 ,  6.25,  5.94,  5.57,  5.6 ],
       [ 5.81,  5.32,  6.42,  5.97,  6.  ],
       [ 6.49,  6.14,  5.81,  5.45,  5.47],
       [ 6.15,  5.76,  5.92,  5.53,  5.56],
       [ 6.08,  5.83,  4.89,  4.61,  4.63],
       [ 5.36,  5.03,  5.07,  4.74,  4.77]])

以上で主成分分析が完了しました．
*****